### PR TITLE
Laboratory work #2, Zheleznova Tatiana - 22FPL2

### PR DESCRIPTION
**Pull request by tanya-iron**

### CODE REVIEW

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,6 +14,15 @@ def prepare_word(
     :param end_of_word: a token that signifies the end of word
     :return: preprocessed word
     """
+    if (not isinstance(raw_word, str)
+            or not isinstance(start_of_word, str | None)
+            or not isinstance(end_of_word, str | None)):
+        return None
+    if start_of_word is None:
+        return tuple([sign for sign in raw_word] + [end_of_word])
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:22]: are you sure you need the for. And why do you think, if `start_of_word` is `None` that `end_of_word` is not?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (tanya-iron) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -14,7 +14,6 @@ def tokenize(text: str) -> list[str] | None:
     """
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:14]: Revert the 1st lab!

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -26,6 +36,20 @@ def collect_frequencies(
...
+    if (not isinstance(text, str)
+            or not isinstance(start_of_word, str | None)
+            or not isinstance(end_of_word, str)):
+        return None
+    frequencies = {}
+    for word in text.split(' '):
+        word_key = prepare_word(word, start_of_word, end_of_word)
+        if word_key is None:
+            return None
+        if word_key in frequencies:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:48]: simplify - remove else (set 0 if the key is not in dict)

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -36,6 +60,17 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    token_pairs = {}
+    for word in word_frequencies:
+        for index, token in enumerate(word):
+            if (index + 1) < len(word):
+                if (token, word[index + 1]) in token_pairs:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:69]: the same

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (tanya-iron) [reply lab_1_classify_by_unigrams/main.py:14]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (tanya-iron) [reply lab_2_tokenize_by_bpe/main.py:48]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (tanya-iron) [reply lab_2_tokenize_by_bpe/main.py:69]: fixed

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -26,6 +36,20 @@ def collect_frequencies(
...
+            or not isinstance(end_of_word, str)):
+        return None
+    frequencies = {}
+    for word in text.split(' '):
+        word_key = prepare_word(word, start_of_word, end_of_word)
+        if word_key is None:
+            return None
+        if word_key not in frequencies:
+            frequencies[word_key] = 0
+        if word_key in frequencies:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:50]: Really need this if?

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -69,6 +135,21 @@ def get_vocabulary(
...
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(unknown_token, str):
+        return None
+    identifier_dict = {}
+    tokens_list = set()
+    for token in word_frequencies:
+        for letter in token:
+            tokens_list.add(letter)
+            for symbol in letter:
+                tokens_list.add(symbol)
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:146]: update

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -81,6 +162,24 @@ def decode(
...
+        return None
+
+    decoded_list = []
+    for number in encoded_text:
+        token_list = [token for token, value in vocabulary.items() if value == number]
+        for token in token_list:
+            if end_of_word_token:
+                if end_of_word_token in token:
+                    deleted_token = len(end_of_word_token)
+                    decoded_list += end_of_word_token[deleted_token:] + ' '
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:177]: strange code. Do you replace `end_of_word_token` with ' '?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
Comment (tanya-iron) [reply lab_2_tokenize_by_bpe/main.py:50]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 7 above)
Comment (tanya-iron) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (tanya-iron) [reply lab_2_tokenize_by_bpe/main.py]: fixed

**Pull request closed without merging**